### ISSUE TITLE
[98] ASAM: Adaptive Sharpness-Aware Minimization for Scale-Invariant Learning of Deep Neural Networks

### ISSUE THREAD

**Issue by dhkim0225** [dhkim0225/1day_1paper#128]
[paper](https://arxiv.org/pdf/2102.11600.pdf) [code](https://github.com/SamsungLabs/ASAM)

역시 이런 류는 icml 이지..
대부분의 sharpness 연구들은 scale dependency problem 이 있다. #126 
요 문제를 해결하는 metric 들이 제안되었는데, ([1](https://arxiv.org/pdf/1711.01530.pdf), [2](https://proceedings.neurips.cc/paper/2019/file/9edda0fd4d983bf975935cfd492fd50b-Paper.pdf), [3](https://proceedings.mlr.press/v119/tsuzuku20a.html), [4](https://arxiv.org/pdf/1903.02237.pdf))
metric들을 제안한 paper들이 직접 학습들을 돌려보고 하지는 않았다.

3개의 contribution.
1. Adaptive Sharpness 제안
    1. stronger correlation with generalization than sharpness does
2. adaptive sharpness를 활용한 learning 방법 제안
    1. weight scale 에 따라 region을 maximization 하는 쪽으로 학습을 시킴
3. 실험적으로 증명

## SAM
SAM(#127) 설명은 생략.
SAM 결론만 살펴보면 다음과 같다.
![image](https://user-images.githubusercontent.com/16400591/194223704-60eaeae4-be49-4260-a2c4-ac06d15d24a8.png)
![image](https://user-images.githubusercontent.com/16400591/194223627-de54635a-0a46-4e52-9ca7-365be3bf1412.png)

## ASAM
A가 loss function을 바꾸지 않는 scaling operator라고 할 때, (#126 참고)
분명 loss contour 등고선 간의 region 이 좁아질 수 있는데, 이럴 때도, $\rho$ 값이 유지되는건 문제가 있다.
![image](https://user-images.githubusercontent.com/16400591/194228678-2bf6a220-6a02-4de8-abdc-a7ea5e0e4d92.png)

우선 Normalization operator 를 정의한다.
${\huge T_w, w \in \mathbb{R}^k}$ 가 invertible linear operator 들이라고 놓자. given weight $w$ 에 대해 loss function을 건들이지 않고 ${\huge T^{-1}_{Aw}A = T^{-1}_w}$ 이라면 ${\huge T^{-1}_w}$ 를 normalization operator 라고 하자.

이를 만족시키는 $T_w$ 는 다음 두개를 써 볼 수 있다.
![image](https://user-images.githubusercontent.com/16400591/194260376-f1cb6aed-6fb1-4350-8c86-7b3546ed5dfd.png)

실제로 $T_w$ 를 계산할 때는, 안정성을 위해 $T_w +ηI_k$ 형태로 작은 $η$ 를 함께 썼다. (0.05)
https://github.com/SamsungLabs/ASAM/blob/f156a680171db16d551c0d85cba2514fa3bff6a2/asam.py#L24

Adaptive sharpness 는 다음과 같다.
![image](https://user-images.githubusercontent.com/16400591/194233422-d98c965f-77cc-49ab-84af-0468a28eccf6.png)
논문에서 증명한 properties 까지 담고 있는 식.
결국, sharpness 값이 transformation 과 무관하게, 일정하게 나오게 만들어 준다는 것이다.

$w = (1, 1)$ 에 scaling operator $A=diag(3, 1/3)$ 을 곱해서 $w^{'} = (3, 1/3)$ 가 나왔다 가정하자.
파란색 line 위의 w 기준으로 SAM 에서처럼 $\max{L(w+\epsilon)}$에 해당하는 $\epsilon$ 값을 구해서 weight 를 구하고, Transforms들을 적용하면서, 빨간 line 을 그린다.
![image](https://user-images.githubusercontent.com/16400591/194228152-b01e96fe-a9e9-4dc0-8af0-eda635b2326e.png)

제안하는 방법은 epsilon ball 에도 transform 이 적용이 된다. (그림 참고)
generalization gap 은  mini-batch size, initial learning rate, weight decay coefficient, dropout rate 4개를 변경해 보면서, generalization gap 을 측정했다.
정의한 adaptive sharpness 와 correleation 을 찍어보면 유의미하다.
![image](https://user-images.githubusercontent.com/16400591/194246790-7a173bf3-21ff-4ec8-b8fd-6e4c4cf3a8ab.png)

## Adaptive Sharpness-Aware Minimization
그래서 minimization 은 어떻게 하냐.
![image](https://user-images.githubusercontent.com/16400591/194248922-dd05a65a-9467-40f4-97b7-a2157b7e3cf2.png)
![image](https://user-images.githubusercontent.com/16400591/194248874-bf65ea46-e33f-4c26-9a27-64672c4ed7f8.png)

## Results
norm 은 L2 쓰는게 좋다. $\rho$ 값은 여전히 hyperparameter.
![image](https://user-images.githubusercontent.com/16400591/194247586-fdd40d9d-9487-4080-8765-c2852eac5db7.png)

cifar-10
![image](https://user-images.githubusercontent.com/16400591/194247923-be6c709c-5ca1-4541-8769-6e9aaf8faac6.png)

cifar-100
![image](https://user-images.githubusercontent.com/16400591/194247976-91afe6c6-86da-4699-95bd-b628d5936a11.png)

Imagenet
![image](https://user-images.githubusercontent.com/16400591/194248059-58ef0066-eeca-48db-afe9-c6cc43230c7f.png)

MNT (DE-EN)
![image](https://user-images.githubusercontent.com/16400591/194248180-2ba6929b-7e6d-4c61-9232-61fdc32b2d9e.png)

SAM 은 라벨 노이즈에 강인하다?
![image](https://user-images.githubusercontent.com/16400591/194248315-eacacb5e-76a1-42f9-ab27-77c4c201ece8.png)
